### PR TITLE
Global setting to select preferred storage pool

### PR DESCRIPTION
**Pull request by ravening**
### Description
Currently all the volumes are allocated on storage pools
based on the capacity or the algorithm selected. Sometimes
we need to deploy all volumes of particular account in a
specific storage pool and in that case its not possible.

with this change, we can specify the uuid of the preferred
storage pool, so that all volumes of the account will be
deployed in this pool

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### Screenshots (if appropriate):
Screenshot 2021-07-28 at 14 01 48

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@ravening, one nit from functional perspective; wouldn't a user want to use a storage pool name instead of a uuid?

**Comment by ravening** [comment]
> @ravening, one nit from functional perspective; wouldn't a user want to use a storage pool name instead of a uuid?

@DaanHoogland This will be configured mostly by admins. Also if you see similar other global settings , they all use uuid's rather than actual name of the resource. For eg: router service offering

**Comment by DaanHoogland** [comment]
> > @ravening, one nit from functional perspective; wouldn't a user want to use a storage pool name instead of a uuid?
> 
> @DaanHoogland This will be configured mostly by admins. Also if you see similar other global settings , they all use uuid's rather than actual name of the resource. For eg: router service offering

Yes but this is a preferred pool for an account. wouldn't the account holders have to/want to configure that themselves? And wouldn't the admin want to not be bothered and leave it to the users?

File: server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java [hunk 1]
```diff
@@ -1674,7 +1675,12 @@ protected boolean hostCanAccessSPool(Host host, StoragePool pool) {
             for (StoragePoolAllocator allocator : _storagePoolAllocators) {
                 final List<StoragePool> suitablePools = allocator.allocateToPool(diskProfile, vmProfile, plan, avoid, returnUpTo);
                 if (suitablePools != null && !suitablePools.isEmpty()) {
-                    suitableVolumeStoragePools.put(toBeCreated, suitablePools);
+                    List<StoragePool> pools = new ArrayList<>();
+                    Optional<StoragePool> storagePool = getPreferredStoragePool(suitablePools, vmProfile.getVirtualMachine());
+                    storagePool.ifPresent(pools::add);
+
+                    pools.addAll(suitablePools);
+                    suitableVolumeStoragePools.put(toBeCreated, pools);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:1683]: can you extract this bit as to not enlarge an already too long method, please?

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 2]
```diff
@@ -321,9 +347,13 @@ public StoragePool findStoragePool(DiskProfile dskCh, DataCenter dc, Pod pod, Lo
...
 
-            final List<StoragePool> poolList = allocator.allocateToPool(dskCh, profile, plan, avoidList, 1);
+            final List<StoragePool> poolList = allocator.allocateToPool(dskCh, profile, plan, avoidList, StoragePoolAllocator.RETURN_UPTO_ALL);
             if (poolList != null && !poolList.isEmpty()) {
-                return (StoragePool)dataStoreMgr.getDataStore(poolList.get(0).getId(), DataStoreRole.Primary);
+                // Check if the preferred storage pool can be used. If yes, use it.
+                Optional<StoragePool> storagePool = getPreferredStoragePool(poolList, vm);
+
+                return (storagePool.isPresent()) ? (StoragePool) this.dataStoreMgr.getDataStore(storagePool.get().getId(), DataStoreRole.Primary) :
+                    (StoragePool)dataStoreMgr.getDataStore(poolList.get(0).getId(), DataStoreRole.Primary);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:356]: can you extract this into a `private StoragePool checkPreferredStorageForUse(..)` and than store the result in a `StoragePool poolToReturn = null;` so we have only one exit point. I.E. end the method with `return poolToReturn`

**Review by DaanHoogland — APPROVED** [review]
clgtm, (some suggestions added) I do have minor functional doubts, but don't think they should hinder this PR (can be a follow up).

**Comment by ravening** [comment]
> > > @ravening, one nit from functional perspective; wouldn't a user want to use a storage pool name instead of a uuid?
> > 
> > 
> > @DaanHoogland This will be configured mostly by admins. Also if you see similar other global settings , they all use uuid's rather than actual name of the resource. For eg: router service offering
> 
> Yes but this is a preferred pool for an account. wouldn't the account holders have to/want to configure that themselves? And wouldn't the admin want to not be bothered and leave it to the users?

@DaanHoogland customers wont have access or see the list of storage pools in the platform. So they cant select it anyway. its useful only for admins.

**Comment by DaanHoogland** [comment]
> > > > @ravening, one nit from functional perspective; wouldn't a user want to use a storage pool name instead of a uuid?
> > > 
> > > 
> > > @DaanHoogland This will be configured mostly by admins. Also if you see similar other global settings , they all use uuid's rather than actual name of the resource. For eg: router service offering
> > 
> > 
> > Yes but this is a preferred pool for an account. wouldn't the account holders have to/want to configure that themselves? And wouldn't the admin want to not be bothered and leave it to the users?
> 
> @DaanHoogland customers wont have access or see the list of storage pools in the platform. So they cant select it anyway. its useful only for admins.

too bad

**Review by GabrielBrascher — APPROVED** [review]
Overall code LGTM, I see that @DaanHoogland made some interesting suggestions. I would appreciate it if you take them into account. But either way, LGTM.

Thanks for the PR @ravening.
Has this been tested? Manual tests might be interesting in this one with some alternatives for the global settings.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by ravening** [comment]
> Overall code LGTM, I see that @DaanHoogland made some interesting suggestions. I would appreciate it if you take them into account. But either way, LGTM.
> 
> Thanks for the PR @ravening.
> Has this been tested? Manual tests might be interesting in this one with some alternatives for the global settings.

@GabrielBrascher will work on it to make changes.

Yes I tested them manually by passing uuid of particular pool in the global setting. All of the vm deployment, volume attachment to a vm will happen on this pool.

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java (hunk 2 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:356]: We can add the availability to the loop condition and break once found, can't we? I see your point though. Though the principle stands, I'm not making a big issue of it.

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by nvazquez — APPROVED** [review]
LGTM

Tested manually on Vmware 6.7 + 2 NFS storage pools:
- Created a user account (account1) and a domain admin account(account2)
- admin) set 'preferred.storage.pool' on account2 to `pool2`
- account2) deployed multiples VMs
- admin) deployed multiple VMs - verify VM volumes are distributed across storage pools and account2 VM volumes are on `pool2`
- account1) deploy multiple VMs
- admin) verify account1 VM volumes are distributed across storage pools
- admin) set 'preferred.storage.pool' on account2 to `pool1`
- account2) deploy multiple VMs
- admin) verify account2 new VM volumes are on `pool1`

**Comment by levindecaro** [comment]
@ravening Is it possible to extend this settings to "Project Account", it seems to be more useful instead of individuals. BTW, if preferred UUID presented in "Global" Setting and "User Account" at the same time, ACS will take which as preference? Thanks

**Pull request merged**